### PR TITLE
ESP update and general cleanup

### PR DESCRIPTION
**Pull request by KeithWoods**
* Updated [esp](https://github.com/esp/esp-js) to vlatest (to get better model to model api)
* Integrated [esp-js-react](https://github.com/esp/esp-js-react)
  * removing ViewBase
  * removing `@view` in favour of `@esp.viewBinding`
  * using `<SmartComponent />` for model to component wire-up
  * using `<RouterProvider />` to expose the router throughout the vDom
* Merging analytics and sidebar region, they are the same thing
* Removing hacks between analytics, sidebar and the shell to aid in showing hiding analytics. Regions should act independently of their content, region hosts (i.e. ShellView) should act independent of the content within the regions.
* Re-worked main flexbox layout so it dynamically sizes based on content without the need for code to swap out class names.
* replaced react-dimensions with react-sizeme

### CODE REVIEW

File: src/client/src/appBootstrapper.js [hunk 1]
```diff
@@ -94,32 +101,22 @@ class AppBootstrapper {
...
     analyticsRegionModel.observeEvents();
-    let sidebarRegionModel = new SingleItemRegionModel(WellKnownModelIds.sidebarRegionModelId, RegionNames.sidebar, espRouter);
-    sidebarRegionModel.observeEvents();
-    let regionManager = new RegionManager(
-      [workspaceRegionModel, popoutRegionModel, blotterRegionModel, analyticsRegionModel, sidebarRegionModel], this._openFin.isRunningInOpenFin);
-
-    // wire up the shell
-    let shellModel = new ShellModel(WellKnownModelIds.shellModelId, espRouter, this._connection, this._openFin);
-    shellModel.observeEvents();
+    var allRegionModels = [workspaceRegionModel, popoutRegionModel, blotterRegionModel, analyticsRegionModel];
```
Comment (raybooysen) [inline src/client/src/appBootstrapper.js]: can be a `let`  minor

Comment (KeithWoods) [inline src/client/src/appBootstrapper.js]: nice catch. I blame intellij's have extract variable :)

File: src/client/src/ui/regions/views/sidebar/sidebarRegionView.jsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    router:React.PropTypes.object.isRequired
+  };
+
+  render() {
+    let model:SingleItemRegionModel = this.props.model;
+    let innerContent = null;
+
+    if (model.modelRegistrations.length === 1) {
+      let className = classnames(
+        'sidebarRegion__content',
```
Comment (fdore) [inline src/client/src/ui/regions/views/sidebar/sidebarRegionView.jsx]: sidebarRegion should be sidebar-region to be consistent

File: src/client/src/ui/analytics/views/analyticsView.jsx [hunk 3]
```diff
@@ -1,26 +1,26 @@
...
 import './analytics.scss';
 
 var _log:logger.Logger = logger.create('AnalyticsView');
 
-@Dimensions()
-export default class AnalyticsView extends ViewBase {
+export default class AnalyticsView extends React.Component {
+
+  static propTypes = {
+    model: React.PropTypes.object.isRequired,
```
Comment (raybooysen) [inline src/client/src/ui/analytics/views/analyticsView.jsx]: can we use `typeof` on propTypes for these?

File: src/client/src/ui/analytics/views/analyticsView.jsx [hunk 4]
```diff
@@ -72,7 +69,7 @@ export default class AnalyticsView extends ViewBase {
       <div className='analytics analytics__container animated fadeIn'>
         <div className='analytics__controls popout__controls'>
           <i className={newWindowBtnClassName}
-             onClick={() => router.publishEvent(this.props.modelId, 'popOutAnalytics', {})}/>
+             onClick={() => router.publishEvent(this.props.model.modelId, 'popOutAnalytics', {})}/>
```
Comment (raybooysen) [inline src/client/src/ui/analytics/views/analyticsView.jsx]: We have the model as a variable above, can we just that instead?

File: src/client/src/ui/blotter/model/blotterModel.js [hunk 5]
```diff
@@ -46,6 +47,10 @@ export default class BlotterModel extends ModelBase {
     this._schedulerService = schedulerService;
   }
 
+  get canPopout() {
+    return Environment.isRunningInIE;
```
Comment (raybooysen) [inline src/client/src/ui/blotter/model/blotterModel.js]: What an odd property name.  :) Do pop-outs only work in IE?

File: src/client/src/ui/blotter/views/blotterView.jsx [hunk 6]
```diff
@@ -1,34 +1,31 @@
...
 import 'fixed-data-table/dist/fixed-data-table.css';
 import './blotter.scss';
 
-@Dimensions()
-export default class BlotterView extends ViewBase {
+class BlotterView extends React.Component {
+
+  static propTypes = {
+    model: React.PropTypes.object.isRequired,
+    router: React.PropTypes.object.isRequired,
```
Comment (raybooysen) [inline src/client/src/ui/blotter/views/blotterView.jsx]: Same here I guess.

File: src/client/src/ui/blotter/views/blotterView.jsx [hunk 7]
```diff
@@ -41,21 +38,20 @@ export default class BlotterView extends ViewBase {
...
-    let containerWidth = this.props.containerWidth; // comes from the @Dimensions annotation
-    let containerHeight = this.props.containerHeight; // comes from the @Dimensions annotation
+    let { width, height } = this.props.size; // comes from SizeMe
     return (
       <div className={className}>
         <div className='blotter-wrapper'>
           <div className='blotter__controls popout__controls'>
             <i className={newWindowClassName}
-               onClick={() => router.publishEvent(this.props.modelId, 'tearOffBlotter', {})} />
+               onClick={() => this.props.router.publishEvent(this.props.model.modelId, 'tearOffBlotter', {})} />
```
Comment (raybooysen) [inline src/client/src/ui/blotter/views/blotterView.jsx]: We have the model above, should use that.  A few instances below

File: src/client/src/ui/blotter/views/mySizeAwareComponent.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  size: PropTypes.shape({
+    width: PropTypes.number.isRequired,
+    height: PropTypes.number.isRequired,
+  }),
+  style: PropTypes.object,
+};
+
+export default sizeMe({
+  monitorHeight: true,
+ // refreshRate: 2500,
```
Comment (raybooysen) [inline src/client/src/ui/blotter/views/mySizeAwareComponent.js]: can be removed?

File: src/client/src/ui/regions/views/sidebar/sidebarRegionView.jsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+      {
+        'sidebar-region__element--active': innerContent !== null && !model.isCollapsed,
+        'sidebar-region__element--inactive' :  innerContent === null || model.isCollapsed
+      }
+    );
+    return (
+      <div className={this.props.className}>
+        {innerContent}
+        <div className='sidebar-region__container'>
+          <i className={buttonClassName} onClick={() => this.props.router.publishEvent(this.props.model.modelId, 'toggleIsCollapsed', {})} />
```
Comment (raybooysen) [inline src/client/src/ui/regions/views/sidebar/sidebarRegionView.jsx]: We have the model above.  A few of the files have this, won't comment on each

File: src/client/src/ui/spotTile/views/spotTileView.jsx [hunk 10]
```diff
@@ -1,33 +1,26 @@
...
 import './spotTile.scss';
 
 const SPOT_DATE_FORMAT = 'DD MMM';
 
-export default class SpotTileView extends ViewBase {
+export default class SpotTileView extends React.Component {
+
+  static propTypes = {
+    model: React.PropTypes.object.isRequired,
+    router: React.PropTypes.object.isRequired
```
Comment (raybooysen) [inline src/client/src/ui/spotTile/views/spotTileView.jsx]: A few of the propTypes in files could use instanceOf or Typeof whichever for the model and router.

File: src/client/src/ui/analytics/views/analyticsView.jsx (hunk 3 above)
Comment (KeithWoods) [inline src/client/src/ui/analytics/views/analyticsView.jsx]: There is some circular dependency issues with the require statement whereby if you require the model here for the type of check it sometimes works and sometimes is undefined (the model also requires the view for view  resolution).  I didn't have time to look into it further.

File: src/client/src/ui/blotter/views/mySizeAwareComponent.js (hunk 8 above)
Comment (KeithWoods) [inline src/client/src/ui/blotter/views/mySizeAwareComponent.js]: oh crap, yes

File: src/client/src/ui/blotter/views/blotterView.jsx (hunk 7 above)
Comment (KeithWoods) [inline src/client/src/ui/blotter/views/blotterView.jsx]: done

File: src/client/src/ui/blotter/views/blotterView.jsx (hunk 6 above)
Comment (KeithWoods) [inline src/client/src/ui/blotter/views/blotterView.jsx]: will do

File: src/client/src/ui/spotTile/views/spotTileView.jsx (hunk 10 above)
Comment (KeithWoods) [inline src/client/src/ui/spotTile/views/spotTileView.jsx]: same reason as before, some circular ref resolution with webpacks require

File: src/client/src/ui/blotter/model/blotterModel.js (hunk 5 above)
Comment (KeithWoods) [inline src/client/src/ui/blotter/model/blotterModel.js]: good god I only noticed this. It was there before my PR and I blindly lifted it from the middle of the file to near the top.

I had a quick search and looks like this prop is in a few place. It's very odd indeed.

@fdore does it ring any bells?

**Pull request merged**